Copy Jobs to batch/v1 group

erictune opened the pull request.

erictune committed:
Created batch API group and copied Job there

Modeled after on first commit (2fbc5bb) of piosz:hpa-ga (#20501).

erictune committed:
Add conversion functions.

This is needed for Job because it uses PodSpec and PodSpec
does not have auto-generatable conversion for some reason.

Copied pkg/apis/extensions/v1beta1/conversion.go, and removed
non-Job things.

erictune committed:
Enabled batch API group in apiserver

erictune committed:
Fixed and added tests

erictune committed:
Implemented Batch client

erictune committed:
Autogenerate.

erictune commented:
@lavalamp this should be working well enough for you to take a look at it.  I'm waiting to see what integration and e2e testing says, but unit tests pass.

lavalamp commented on hack/verify-api-reference-docs.sh:
```diff
@@ -35,7 +35,7 @@ TMP_ROOT="${KUBE_ROOT}/_tmp"
 echo "diffing ${API_REFERENCE_DOCS_ROOT} against freshly generated docs"
 ret=0
-diff -NauprBZ -I 'Last update' --exclude=*.md "${API_REFERENCE_DOCS_ROOT}" "${OUTPUT_DIR}" || ret=$?
+diff -NauprB -I 'Last update' --exclude=*.md "${API_REFERENCE_DOCS_ROOT}" "${OUTPUT_DIR}" || ret=$?
```
Is this diff intentional? seems unrelated?

erictune replied:
I can't regenerate on my machine without this.  But I can take it out of this PR.

lavalamp commented on pkg/api/testapi/testapi.go:
```diff
@@ -78,6 +81,12 @@ func init() {
+	if _, ok := Groups[batch.GroupName]; !ok {
+		Groups[batch.GroupName] = TestGroup{
+			externalGroupVersion: unversioned.GroupVersion{Group: batch.GroupName, Version: registered.GroupOrDie(batch.GroupName).GroupVersion.Version},
+			internalGroupVersion: extensions.SchemeGroupVersion,
```
I missed this in @piosz's change-- can you move the following `if _, ok := Groups[extensions.GroupName]; !ok {` section above the batch & autoscaling setup code? otherwise, referencing the extensions group here may not do what we want.

erictune replied:
ok

lavalamp commented:
No major comments.

erictune committed:
Fix test and reorder initialization of groups

Delete a job scale test
A subsequent PR is going to remove support
for this anyways.

Initialize extensions before batch and autoscaling
per @lavalamp review suggestion.

erictune commented:
Both review comments addressed.  Also, commented out a scaler test which was failing: I will remove Scaler from Job in a subsequent PR, as already requested by bgrant0607

erictune commented:
@k8s-bot please unit test again, flake #21451

erictune commented:
@lavalamp review comments addressed and passing all tests except for a flake on the last run.  Ready for LGTM.

erictune committed:
Fix Job Reaping

Ensure batch.Kind("Job") has a reaper, so that pods are not orphaned.
Check for orphaned pods in test-cmd.sh.
Also provide describer and scaler for batch.Kind("Job").
The scaler, reaper, and describer for extensions can
be reused for batch.

erictune commented:
Found problem in kubect, pushed fix.

soltysh commented:
LGTM

erictune committed:
Comment out test to fix later.

erictune commented:
I intend the commit named "Comment out test to fix later" to merge.

erictune committed:
Fix jobs integration test.

erictune commented:
Applying LGTM that @lavalamp gave over chat.

erictune commented:
@k8s-bot e2e test this issue #21463

erictune commented:
@k8s-bot e2e test this issue #21463

lavalamp commented:
\o/

On Mon, Feb 22, 2016 at 12:13 AM, k8s-merge-robot <notifications@github.com>
wrote:

> Automatic merge from submit-queue
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21434#issuecomment-187063813>
> .
>
